src/node_contextify.cc

```diff
@@ -270,6 +271,83 @@ void ContextifyContext::IsContext(const FunctionCallbackInfo<Value>& args) {
 }
 
 
+void ContextifyContext::GetWrappedFunction(
+    const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Local<Context> context = env->context();
+  Isolate* isolate = env->isolate();
+
+  CHECK(args[0]->IsString());
+  Local<String> code = args[0].As<String>();
+
+  CHECK(args[1]->IsBoolean());
+  bool produce_cached_data = args[1]->IsTrue();
+
+  Local<Uint8Array> cached_data_buf;
+  if (!args[2]->IsUndefined()) {
+    CHECK(args[2]->IsUint8Array());
+    cached_data_buf = args[2].As<Uint8Array>();
+  }
+
+  // TODO(ryzokuken): Add support for sandboxing.
+
+  ScriptCompiler::CachedData* cached_data = nullptr;
+  if (!cached_data_buf.IsEmpty()) {
+    ArrayBuffer::Contents contents = cached_data_buf->Buffer()->GetContents();
+    uint8_t* data = static_cast<uint8_t*>(contents.Data());
+    cached_data = new ScriptCompiler::CachedData(
+      data + cached_data_buf->ByteOffset(), cached_data_buf->ByteLength());
+  }
+
+  ScriptCompiler::Source source(code, cached_data);
+
+  ScriptCompiler::CompileOptions options;
+  if (source.GetCachedData() == nullptr) {
+    options = ScriptCompiler::kNoCompileOptions;
+  } else {
+    options = ScriptCompiler::kConsumeCodeCache;
+  }
+
+  Local<String> params[] = {
+    env->module_parameter_exports(),
+    env->module_parameter_require(),
+    env->module_parameter_module(),
+    env->module_parameter_filename(),
+    env->module_parameter_dirname()
+  };
```

Comment:
It might actually be nice if we could pass these in as arguments to for `GetWrappedFunction()`?

Reply:
That would be nice, yes. That way this function only provides a primitive to build upon.

Reply:
> do I pass these into the primitive function which requires them?

Yes, that. (Assuming “requires” = “accepts as arguments”)

Reply:
I think an empty array would make a sensible default here

Reply:
@addaleax cool. So should I add the said primitive in this PR, or should I follow up with a PR that refactors most of the current code into the primitive, with extra code to handle params as args to the new function would just call the primitive with these params as an arg?

Reply:
not sure which parts you are referring to by “current code”, but generally, you can use a single PR that contains two logically separate commits:

- One that adds the CompileFunctionInContext binding to the vm module and does not touch the module load
- One that makes the CJS loader use that binding and doesn’t touch the VM module

Does that help?

Reply:
It does.